Add a corpus structure editor (#22)
hexatomic/hexatomic: docs/user/src/usage/corpus-structure.md

# Editing the corpus structure

Each corpus project can consist of multiple documents which are organized into

- corpus graphs,
- corpora, and
- sub-corpora.

> Even in simple projects with only one document, all this corpus structure exists and can be used to extend or re-organize existing corpora.

Reviewer: ```suggestion
Even in simple projects with only one document, this corpus structure exists and can be used to extend or re-organize existing corpora.
The minimal structure for a corpus with one document is therefore: One corpus graph containing one corpus containing one document.
```

> In Hexatomic, the corpus structure is always visible in the special “Corpus Structure” editor.

Reviewer: This suggests that there will be no other tools for corpus management. Which is probably fine for now, but should be kept in mind.

Reviewer: Also, what happens if the user closes the editor?

Author: > Also, what happens if the user closes the editor?

The Part Window is explicitly marked as non closable and there is no close button.

Reviewer: Can still be closed by moving the editor outside of the window, then closing it via the window control. Anyway, I'd just do this:

```suggestion
In Hexatomic, the “Corpus Structure” editor shows the current corpus structure and can be used to make changes to it.
```

![An example corpus structure](./example-corpus-structure.png)

## Corpus graphs

> Corpora, sub-corpora and documents are organized in hierarchies, the so-called corpus graphs.

Reviewer: ```suggestion
Corpora, sub-corpora and documents are organized in a hierarchical structure, the so-called corpus graph.
```

> A project in Hexatomic *can* have more than one corpus-graphs, but for most projects a single corpus graph is sufficient.

Reviewer: ```suggestion
A project in Hexatomic *can* have more than one corpus-graph, but for most projects a single corpus graph is sufficient.
```

In the special case where you import different corpora from different annotation formats into the same Hexatomic project for merging them, you will need more than one corpus graph.

In an empty project, just click on the “Add“ button to add a new corpus structure.

![Add button for default action](./corpus-structure-add-default.png)

> Just using the default “Add“ button is context sensitive depending on which element is selected in the corpus structure.

Reviewer: ```suggestion
The default “Add“ button is context-sensitive, and will add elements "intelligently", depending on which type of element is currently selected in the corpus structure.
```

To explicitly choose the element to add, click on the small arrow on the right side of the button and a drop-down menu with the different options will appear.

![Add button for specific action](./corpus-structure-add-specific.png)

> If you delete a corpus graph, all of its documents and corpora will also be deleted.

Reviewer: What if you delete a corpus, will all sub-corpora and documents also be deleted?

Author: Currently whatever Salt does on 
`corpus.getGraph().removeNode(corpus)`
It seems to be attached to the parent node, but we might want to fix Salt if we don't like this behavior. I disallowed deleting (sub-) corpora with children and documented it in 1288e8656d7a6e4a054c8ae1dc14ea4155777e15

Reviewer: Works for me.

## Corpora and sub-corpora

Inside a corpus graph, the different corpora and sub-corpora are organized as a hierarchy.

> A corpus graph should only contain one toplevel corpus which name is often used as corpus name when exporting the corpus to a different format.

Reviewer: ```suggestion
A corpus graph should only contain one top-level corpus, whose name is often used as corpus name when exporting the corpus to a different format.
```

Bit weird, I know, but things are also who in this case...

To add a sub-corpus, select the parent corpus, click on the arrow on the right side of the “Add” button and choose “(Sub-) Corpus”.
You can edit the name of a corpus by double-clicking on its entry and pressing enter when finished.

![Rename a corpus](./corpus-structure-rename.png)

## Documents

> When a corpus is selected, the default action for the “Add” button is to a a new document.

Reviewer: ```suggestion
When a corpus is selected, the default action for the "Add" button is to add a new document.
When a document is selected, the "Add" button will create a new sibling document in the same parent corpus.
```

> Documents must have a corpus as a parent and contain the linguistic annotations.

Reviewer: ```suggestion
Documents must have a corpus as a parent and contain the base text and linguistic annotations.
```

Or find a better way to say this without making it too complicated for the user, who might wonder where the base data source will be if documents only contain annotations...

You can move a document from one (sub-) corpus to another by dragging and dropping it.

![Drag document](./corpus-structure-drag.png)
![Drop document](./corpus-structure-drop.png)
![Drop document result](./corpus-structure-drop-result.png)

> It is possible to apply a filter, to only show documents which names contain a certain string.

Reviewer: ```suggestion
It is possible to apply a filter, to only show documents whose names contain a certain string.
```
